File: frontend/src/api/chatassistant.ts
import { api } from "@/api/client";
import type { FlowEdge, FlowNode } from "@/types";

export interface ChatResponse {
  reply: string;
  intent: string;
  data: Record<string, unknown> | null;
}

export async function sendChatMessage(
  message: string,
  nodes: FlowNode[],
  edges: FlowEdge[]
): Promise<ChatResponse> {
  const { data } = await api.post<ChatResponse>("/assistant/chat", {
    message,
    nodes: nodes.map((n) => ({ id: n.id, type: n.type, text: n.text, imageUrl: n.imageUrl ?? null })),
    edges: edges.map((e) => ({ id: e.id, fromNodeId: e.fromNodeId, toNodeId: e.toNodeId, label: e.label ?? null })),
  });
  return data;
}

File: frontend/api/app/api/routes/chatassistant.py
from fastapi import APIRouter, Depends

from app.api.deps import get_current_user
from app.chatassistant.pipeline import handle_chat_message
from app.chatassistant.schemas import ChatRequest, ChatResponse
from app.explainer.providers import TextProviderError, get_configured_text_provider
from app.models.user import User

router = APIRouter(prefix="/assistant", tags=["assistant"])


@router.post("/chat", response_model=ChatResponse)
def chat(payload: ChatRequest, _: User = Depends(get_current_user)):
    try:
        provider = get_configured_text_provider()
    except TextProviderError:
        provider = None

    try:
        result = handle_chat_message(
            payload.message,
            [n.model_dump() for n in payload.nodes],
            [e.model_dump() for e in payload.edges],
            provider,
        )
    except Exception as e:
        from fastapi import HTTPException
        err_str = str(e)
        if "quota" in err_str.lower() or "429" in err_str:
            raise HTTPException(status_code=429, detail="Gemini AI API rate limit reached (Quota Exceeded). Please wait a minute and try again!")
        raise HTTPException(status_code=502, detail=f"AI Chat Assistant failed: {e}")

    return ChatResponse(reply=result.reply, intent=result.intent, data=result.data)
